src/relay/ir/Relay.g4

```diff
@@ -0,0 +1,140 @@
+grammar Relay;
+
+// Lexing
+// comments
+WS : [ \t\n\r]+ -> skip ;
+LINE_COMMENT : '//' .*? '\n' -> skip ;
+COMMENT : '/*' .*? '*/' -> skip ;
+
+// operators
+MUL: '*' ;
+DIV: '/' ;
+ADD: '+' ;
+SUB: '-' ;
+LT: '<' ;
+GT: '>' ;
+LE: '<=' ;
+GE: '>=' ;
+EQ: '==' ;
+NE: '!=' ;
+
+operatorIdent: CNAME ;
+GLOBAL_IDENT: '@' CNAME ;
+LOCAL_IDENT: '%' CNAME ;
+
+MUT: 'mut' ;
+
+BOOL_LIT
+  : 'true'
+  | 'false'
+  ;
+
+// non-negative floats
+FLOAT
+  : INT '.' INT EXP? // 1.35, 1.35E-9, 0.3, 4.5
+  | INT EXP // 1e10 3e4
+  ;
+
+// non-negative ints
+// INT: '0' | [1-9] DIGIT* ; // no leading zeros
+INT: DIGIT+ ;
+fragment EXP: [eE] [+\-]? INT ; // \- since - means "range" inside [...]
+
+CNAME: ('_'|LETTER) ('_'|LETTER|DIGIT)* ;
+fragment LETTER: [a-zA-Z] ;
+fragment DIGIT: [0-9] ;
+
+// Parsing
+
+// a program is a list of options followed by either an expression or a list of definitions
+// TODO: is this really the desired outline? if there are defns should there be an expression at the end?
+prog: option* (expr | defn+) EOF ;
+
+option: 'set' ident BOOL_LIT ;
+
+expr
+  // operators
+  // todo: this precedence seems fishy
+  : '(' expr ')'                  # parens
+  | '-' expr                              # neg
+  | expr op=('*'|'/') expr                # binOp
+  | expr op=('+'|'-') expr                # binOp
+  | expr op=('<'|'>'|'<='|'>=') expr      # binOp
+  | expr op=('=='|'!=') expr              # binOp
+
+  // function definition and application
+  | expr '(' (expr (',' expr)*)? ')'      # call
```

Comment:
I feel like we need the function name as an indent instead of expr since expr will accept too much.

Reply:
+1, Also, I can't find a rule for ident expr

Reply:
I kind of agree; however, that becomes restrictive if we want to use curried functions. It's possible to restrict what expressions can be used in calls in the parser itself.

Reply:
If we want to control the form of the program we should not do it in the parser as it makes the parser overly restrictive and forces semantic information into the parser instead of later analysis phases. 

It also disallows programs which are semantically equivalent.

For example:

```
let x = f(x, y, z);
x(z, y)
```

Instead of:
```
 f(x, y, z)(z, y)
```

Even if it looks ugly we should let people generate the text format how they see fit imo.

For example a simple compiler targeting Relay might generate code like:

```
(fn (...) { ... })(x, y, z)
```